core/modules/qproc/IndexMap.cc

```diff
@@ -0,0 +1,252 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+/**
+  * @file
+  *
+  * @brief Implementation of IndexMap
+  *
+  * @author Daniel L. Wang, SLAC
+  */
+
+#include "qproc/IndexMap.h"
+
+// System headers
+#include <algorithm>
+#include <cassert>
+#include <iterator>
+#include <stdexcept>
+#include <set>
+#include <vector>
+
+// Third-party headers
+#include "boost/lexical_cast.hpp"
+
+// LSST headers
+#include "sg/Chunker.h"
+
+// Qserv headers
+#include "global/intTypes.h"
+#include "global/stringTypes.h"
+#include "qproc/geomAdapter.h"
+#include "qproc/SecondaryIndex.h"
+#include "query/Constraint.h"
+
+using lsst::qserv::StringVector;
+using lsst::sg::Region;
+using lsst::sg::Box;
+using lsst::sg::Circle;
+using lsst::sg::Ellipse;
+using lsst::sg::ConvexPolygon;
+using lsst::sg::SubChunks;
+using lsst::qserv::qproc::Coordinate;
+
+typedef std::vector<SubChunks> SubChunksVector;
+
+namespace { // File-scope helpers
+template <typename T>
+std::vector<T> convertVec(StringVector const& v) {
+    std::vector<T> out;
+    std::transform(v.begin(), v.end(), std::back_inserter(out),
+                   boost::lexical_cast<T, std::string>);
+    return out;
+}
+template <typename T>
+boost::shared_ptr<Region> make(StringVector const& v) {
+        return boost::shared_ptr<Region>(new T(convertVec<Coordinate>(v)));
+}
+template <>
+boost::shared_ptr<Region> make<Box>(StringVector const& v) {
+    return lsst::qserv::qproc::getBoxFromParams(convertVec<Coordinate>(v));
+}
+template <>
+boost::shared_ptr<Region> make<Circle>(StringVector const& v) {
+    return lsst::qserv::qproc::getCircleFromParams(convertVec<Coordinate>(v));
+}
+template <>
+boost::shared_ptr<Region> make<Ellipse>(StringVector const& v) {
+    return lsst::qserv::qproc::getEllipseFromParams(convertVec<Coordinate>(v));
+}
+template <>
+boost::shared_ptr<Region> make<ConvexPolygon>(StringVector const& v) {
+    return lsst::qserv::qproc::getConvexPolyFromParams(convertVec<Coordinate>(v));
+}
+
+typedef boost::shared_ptr<Region>(*MakeFunc)(StringVector const& v);
+
+struct FuncMap {
+    FuncMap() {
+        fMap["box"] = make<Box>;
+        fMap["circle"] = make<Circle>;
+        fMap["ellipse"] = make<Ellipse>;
+        fMap["poly"] = make<ConvexPolygon>;
+        fMap["qserv_areaspec_box"] = make<Box>;
+        fMap["qserv_areaspec_circle"] = make<Circle>;
+        fMap["qserv_areaspec_ellipse"] = make<Ellipse>;
+        fMap["qserv_areaspec_poly"] = make<ConvexPolygon>;
+    }
+    typedef std::map<std::string, MakeFunc> Map;
+    Map fMap;
+};
+
+static FuncMap funcMap;
```

Comment:
Sorry to be unclear - I meant the moral equivalent of a switch (just a bunch of if/else clauses).